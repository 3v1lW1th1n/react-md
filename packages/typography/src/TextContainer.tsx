import React, {
  forwardRef,
  FunctionComponent,
  HTMLAttributes,
  ReactElement,
  ReactType,
  ReactNode,
} from "react";
import cn from "classnames";
import { WithForwardedRef } from "@react-md/utils";

/**
 * A union of the available text container sizes. One of these values must be chosen
 * to help set the max width for text.
 */
export type TextContainerSize = "auto" | "mobile" | "desktop";

/**
 * A type describing the text container's children render function. It provides an object containing
 * the correct (and merged) className and exects a renderable element to be returned.
 */
export type TextContainerRenderFunction = (props: {
  className: string;
}) => ReactElement<any>;

/**
 * The base props for rendering the text component.
 */
export interface TextContainerProps extends HTMLAttributes<HTMLDivElement> {
  /**
   * An optional className to merge with typography text container styles.
   */
  className?: string;

  /**
   * The size for the text container. This can usually just be left at the default of `"auto"` since
   * it will automatically transition between `"mobile"` and `"desktop"` based on media queries.
   * However, you can also manually specify `"mobile"` or `"desktop"` if needed.
   */
  size?: TextContainerSize;

  /**
   * The component to render as. By default this will just be a div, but anything can be provided.
   */
  component?: ReactType;

  /**
   * Either a child render function or a react node. If this is not the child render function, a
   * different wrapper component can be provided using the `component` prop.
   */
  children?: ReactNode | TextContainerRenderFunction;
}

type WithRef = WithForwardedRef<HTMLDivElement | ReactType>;
type DefaultProps = Required<Pick<TextContainerProps, "size" | "component">>;
type WithDefaultProps = TextContainerProps & DefaultProps & WithRef;

const TextContainer: FunctionComponent<
  TextContainerProps & WithRef
> = providedProps => {
  const {
    className: propClassName,
    component: Component,
    size,
    children,
    forwardedRef,
    ...props
  } = providedProps as WithDefaultProps;

  const className = cn(
    `rmd-text-container rmd-text-container--${size}`,
    propClassName
  );
  if (typeof children === "function") {
    return (children as TextContainerRenderFunction)({ className });
  }

  return (
    <Component {...props} className={className} ref={forwardedRef}>
      {children}
    </Component>
  );
};

const defaultProps: DefaultProps = {
  size: "auto",
  component: "div",
};

TextContainer.defaultProps = defaultProps;

if (process.env.NODE_ENV !== "production") {
  TextContainer.displayName = "TextContainer";

  let PropTypes = null;
  try {
    PropTypes = require("prop-types");
  } catch (e) {}

  if (PropTypes) {
    TextContainer.propTypes = {
      className: PropTypes.string,
      size: PropTypes.oneOf(["auto", "mobile", "desktop"]),
      component: PropTypes.oneOfType([
        PropTypes.string,
        PropTypes.func,
        PropTypes.object,
      ]),
      children: PropTypes.oneOfType([PropTypes.node, PropTypes.func]),
    };
  }
}

export default forwardRef<HTMLDivElement | ReactType, TextContainerProps>(
  (props, ref) => <TextContainer {...props} forwardedRef={ref} />
);
