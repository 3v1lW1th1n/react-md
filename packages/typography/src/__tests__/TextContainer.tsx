import * as React from "react";
import { create } from "react-test-renderer";
import { mount } from "enzyme";

import TextContainer from "../TextContainer";

describe("TextContainer", () => {
  it("should render as a div by default", () => {
    const container = mount(<TextContainer />);

    expect(container.find("div").length).toBe(1);
  });

  it("should be able to render as a string component prop", () => {
    const component = "section";
    const container = mount(<TextContainer component={component} />);

    expect(container.find(component).length).toBe(1);
  });

  it("should be able to render as a functional component", () => {
    const Component = ({ className, children }: any) => (
      <section className={`${className} custom`}>{children}</section>
    );

    const container = mount(
      <TextContainer component={Component}>Hello, world!</TextContainer>
    );
    expect(container.find(Component).length).toBe(1);
    expect(container.render()).toMatchSnapshot();
  });

  it("should be able to render as a class component", () => {
    class Component extends React.Component<any> {
      public render() {
        const { children, className } = this.props;
        return <section className={`${className} custom`}>{children}</section>;
      }
    }

    const container = mount(
      <TextContainer component={Component}>Hello, world!</TextContainer>
    );
    expect(container.find(Component).length).toBe(1);
    expect(container.render()).toMatchSnapshot();
  });

  it("should default to the auto suffix for the rmd-text-container className", () => {
    const container = mount(<TextContainer />);

    expect(container.find("div").hasClass("rmd-text-container")).toBe(true);
    expect(container.find("div").hasClass("rmd-text-container--auto")).toBe(
      true
    );

    container.setProps({ size: "mobile" });
    expect(container.find("div").hasClass("rmd-text-container")).toBe(true);
    expect(container.find("div").hasClass("rmd-text-container--mobile")).toBe(
      true
    );

    container.setProps({ size: "desktop" });
    expect(container.find("div").hasClass("rmd-text-container")).toBe(true);
    expect(container.find("div").hasClass("rmd-text-container--desktop")).toBe(
      true
    );
  });

  it("should correctly merge the prop className", () => {
    const className = "test-class-name";
    const container = mount(<TextContainer className={className} />);

    expect(container.find("div").hasClass("rmd-text-container")).toBe(true);
    expect(container.find("div").hasClass("rmd-text-container--auto")).toBe(
      true
    );
    expect(container.find("div").hasClass(className)).toBe(true);
  });

  it("should be able to use a children render function", () => {
    expect(
      create(
        <TextContainer>
          {({ className }) => (
            <span className={`${className} span`}>Content</span>
          )}
        </TextContainer>
      ).toJSON()
    ).toMatchSnapshot();
  });
});
