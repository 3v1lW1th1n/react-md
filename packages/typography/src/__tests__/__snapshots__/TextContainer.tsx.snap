// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextContainer should be able to render as a class component 1`] = `
<section
  class="rmd-text-container rmd-text-container--auto custom"
>
  Hello, world!
</section>
`;

exports[`TextContainer should be able to render as a functional component 1`] = `
<section
  class="rmd-text-container rmd-text-container--auto custom"
>
  Hello, world!
</section>
`;

exports[`TextContainer should be able to use a children render function 1`] = `
<span
  className="rmd-text-container rmd-text-container--auto span"
>
  Content
</span>
`;
